packages/server/src/config.ts

```diff
@@ -38,6 +40,17 @@ function databaseHostFromEnv(runningInDocker: boolean, env: EnvVariables): strin
 	return null;
 }
 
+function databaseSSLConfigFromEnv(env: EnvVariables): object {
+	if (env.POSTGRES_SSL_ENABLE || false) {
+		if (env.POSTGRES_SSL_VERIFY === true) {
+			return { rejectUnauthorized: true }
+		} else {
+			return { rejectUnauthorized: false }
+		}
+	}
+	return null;
+}
```

Comment:
I don't understand how this can add support for TLS? It just looks like it makes it ignore errors or not. So does it mean TLS already works by default, but you want to add a way to skip TLS errors?

Reply:
The ssl parameter can either be a boolean (false or true), or an object. If it is an object, it is considered enabled and in that case I want to either choose to verify the certificate of the db or not.

If POSTGRES_SSL_ENABLE is set to false or unset, ssl will be `null` and the connection is established without TLS

Reply:
Thanks for clarifying, that makes sense. Couldn't we just hardcode `ssl: { rejectUnauthorized: false }` and get rid of the configuration variables? It would work for both secure and non-secure connections. It won't work for invalid certificate but presumably if someone is using this they should ensure they have a valid certificate and that it's installed on the system?

Reply:
I guess it depends on your security model. I think the variable is useful and now it provides a sensible default that can be overriden if required. I would prefer to keep it as is, if it is okay with you.

Thank you for taking the time to review this:)

Reply:
Also:

`ssl: { rejectUnauthorized: false }` would make any certificate pass, not error out on invalid

Reply:
> ssl: { rejectUnauthorized: false } would make any certificate pass, not error out on invalid

That doesn't make much sense and would even be a major security issue in PostgresSQL. Are you sure it's not just the config that should be set differently?

Reply:
My logic has been a bit flawed there. I tried to fix the logic with sane default values in mind here: https://github.com/laurent22/joplin/pull/4867/commits/edbd1535e5e31ed1789b947b7beb605452e9a794

`ssl: { rejectUnauthorized: false }` will make it so that Joplin server connects to a PostgreSQL server with any certificate verification

`ssl: { rejectUnauthorized: true}' is the default value, which means the certificate will be verified (in case of self-signed, this returns: `db: Could not connect. Will try again. self signed certificate`

If `POSTGRES_SSL_ENABLE` is unset or set to `false`, SSL will be set to null and the connection to the DB will be made without any encryption.

The drawback of the latest fix is that the values aren't booleans, but have to quite literally be "false" or "true" - not sure how big of a problem this is.

Reply:
Where is the official reference for these config settings?

Reply:
I have followed the following: https://github.com/brianc/node-postgres/tree/master/packages/pg-connection-string

Not sure if that is official or not

Reply:
I think we should simply give access to the query part of the Postgres connection string. That way you can set any additional parameter you want, including SSL, but also CA, cert, SSL mode, client encoding, etc. Otherwise we will need to add many env variables to support all this and it will be cumbersome.

So for example: `POSGRES_QUERY=sslmode=no-verify& sslcert=...&encoding=utf8`, etc.

Then when setting up the pg driver we parse this string and pass it to the config.

Reply:
That sounds smart. I will attempt to implement it in the following days.

Reply:
Just a quick question, how would we go about the parsing?

I have thought about a method such as doing something like this, reading the [0] element of each equalSplit for the key and taking [1] as the value:

```
var str = "sslmode=no-verify&sslcert=hello";

var splitted = str.split("&");

for (i in splitted) {
    var equalSplit = splitted[i].split("=")
    console.log(equalSplit)
}
```

Would you consider this effective for the parsing or would you do it in another way?

Reply:
For parsing, we should use the node-posgres package so that we don't need to maintain custom parsing code. We already include this package (called "pg").

I believe you could simply use the function you've linked to above. If it doesn't allow you to parse only the query, you could fake the rest of the connection string - eg. `parse('postgres://fakeuser@fakehost/fakedb?' + queryToParse)`

Reply:
That sounds good. I will look into it

Reply:
I have done some initial work on this, but I am puzzled with the outcome of the code I have written, especially since it behaves differently in different situations.

In a bare file with a following .ts source code everything works as expected:

```
var parse = require('pg-connection-string').parse;
var queryToParse = "sslmode=no-verify";
var config = parse('postgres://fakeuser@fakehost/fakedb?' + queryToParse);
console.log(config);
```

returns:

```
tsc test.ts && node test.js
[Object: null prototype] {
  sslmode: 'no-verify',
  user: 'fakeuser',
  password: '',
  port: null,
  host: 'fakehost',
  database: 'fakedb',
  ssl: { rejectUnauthorized: false }
}
```

However, when I try to implement the same in the project, ssl is not defined:

```
function PgConnectionStringFromEnv(env: EnvVariables): object {
	var parse = require('pg-connection-string').parse;
	var config = parse('postgres://fakeuser@fakehost/fakedb?' + env.POSTGRES_CONNECTION_STRING);
        console.log(config)
	return {
		ssl: config.ssl || false,
	};
}
```

console.log line output:
```
2021-05-02T17:29:45.936396127Z [Object: null prototype] {
2021-05-02T17:29:45.936412249Z   sslmode: 'disable',
2021-05-02T17:29:45.936416363Z   user: 'fakeuser',
2021-05-02T17:29:45.936418891Z   password: '',
2021-05-02T17:29:45.936421285Z   port: null,
2021-05-02T17:29:45.936423636Z   host: 'fakehost',
2021-05-02T17:29:45.936426019Z   database: 'fakedb'
2021-05-02T17:29:45.936428313Z }
```

Any idea why this could be happening?

Reply:
is `env.POSTGRES_CONNECTION_STRING` actually defined?

Reply:
Yes - I gave a differentiating example, but in the console.log output, you can see `sslmode` set to `disable` without the ssl: parameter being set in the config. 

This goes against what is described at https://github.com/brianc/node-postgres/tree/master/packages/pg-connection- string#tcp-connections - ssl should be set to `false` if `sslmode` is set to `disable`

Reply:
Maybe some different package version. Do we need to update the "pg" package?

Reply:
> Maybe some different package version. Do we need to update the "pg" package?

I don't think so - from the changelog, there have been no changes in handling this and I have tested 8.6 as well without success